syntax = "proto3";
package blackbird.proto;

message UcxRegion {
  bytes worker_address = 1;  // UCX worker address as raw bytes
  uint64 remote_key = 2;     // UCX remote key
  uint64 remote_addr = 3;    // Remote memory address
  uint64 size = 4;           // Buffer size
}

enum MemoryKind {
  MEMORY_KIND_UNSPECIFIED = 0;
  MEMORY_KIND_CPU_DRAM = 1;
  MEMORY_KIND_GPU_HBM = 2;
}

message MemoryPlacement {
  MemoryKind kind = 1;                     // CPU or GPU memory
  repeated UcxRegion regions = 2;          // RDMA-accessible regions
  int32 device_id = 3;                     // GPU device id if GPU memory, else -1
  int32 numa_node = 4;                     // NUMA node for CPU memory, else -1
}

enum DiskKind {
  DISK_KIND_UNSPECIFIED = 0;
  DISK_KIND_NVME = 1;
  DISK_KIND_SSD = 2;
  DISK_KIND_HDD = 3;
}

message DiskPlacement {
  DiskKind kind = 1;           // NVMe/SSD/HDD
  string mount_path = 2;       // mount point or base path
  uint64 capacity = 3;         // bytes reserved for this placement
  string node_id = 4;          // node hosting this disk placement
}

enum WorkerStatus {
  WORKER_STATUS_UNDEFINED = 0;
  WORKER_STATUS_ALLOCATED = 1;
  WORKER_STATUS_WRITING = 2;
  WORKER_STATUS_COMPLETE = 3;
  WORKER_STATUS_FAILED = 4;
  WORKER_STATUS_REMOVED = 5;
}

message WorkerPlacement {
  oneof storage {
    MemoryPlacement memory = 1;
    DiskPlacement disk = 2;
  }
  WorkerStatus status = 3;
  string node_id = 4;
} 